torch/csrc/jit/script/compiler.cpp

```diff
@@ -899,8 +899,13 @@ struct to_ir {
     // Register outputs in each block
     for (const auto& x : mutated_variables) {
       auto tv = save_true->getVar(x, stmt.range());
-      true_block->registerOutput(tv);
       auto fv = save_false->getVar(x, stmt.range());
+      if (!tv->type()->isSubtypeOf(unshapedType(fv->type()))) {
```

Comment:
Consider that the false branch produces a value of type `ft` and the true branch produces a value of type `tt`, which is different than `ft`, but happens to be a subtype of it. This conditional will accept it, but if someone were to negate the condition and swap the branches, then it would actually get rejected. This is counterintuitive and this shouldn't happen, and I see two ways to fix it:
1. Check for type equality (very strict, very easy, should work well for now)
2. Have a type unification function. For example you can unify `Dynamic` with `Float(2, 50)`, because the other one is a refinement of the first one. But it is a commutative property, so `Float(2, 50)` also has to unify with `Dynamic`. Considering the simplicity of our type system it's not hard to write a function that verifies whether two types can be unified, but it might be an overkill at this point.

Reply:
What would be the case that ft->subtype(tt) but !tt->subtype(ft)? The types of tensors are erased with unshapedType, and as far as scalars go, NumberType isn't instantiable correct?

Reply:
For 2: I assume that we would want floats & ints to be returned as a scalar right ?